compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2617,6 +2626,26 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn parse_match_guard_cond(&mut self) -> PResult<'a, P<Expr>> {
+        let snapshot = self.clone();
```

Comment:
The only issue I'm seeing here is that this pessimises the parsing of *every* match arm. In this case I think it is reasonable to instead catch the problem in line 2499 and only in the error arm try to conditionally parse the match arm body. Probably also do that for the substitution cases handled right below.